mitiq/folding_cirq.py

```diff
@@ -339,26 +339,51 @@ def fold_method(circuit: Circuit, stretch: float, **kwargs):
-    """Applies global unitary folding and a final partial folding of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+def fold_global(circuit: Circuit,
+                stretch: float,
+                finish_fold_method: Callable[[Circuit, float, Tuple[Any]], Circuit] = fold_gates_from_left,
+                finish_fold_method_args: Tuple[Any] = ()
+                ) -> Circuit:
+    """Returns a folded circuit obtained by folding the global unitary of the input circuit.
 
-    if not (stretch >= 1):
-        raise ValueError("The stretch factor must be a real number >= 1.")
+    A single unitary fold corresponds to a stretch of 3. If stretch (mod 3) > 1, a local folding method is called
+    until the desired stretch factor is reached.
 
-    # determine the number of integer foldings and the final fractional_stretch
-    num_foldings, fractional_stretch = divmod(stretch - 1, 2)
+    The returned folded circuit has a number of gates approximately equal to stretch * len(circuit).
 
-    # integer circuit folding
-    eye = Circuit()
-    for _ in range(int(num_foldings)):
-        eye += inverse(circuit) + circuit
 
-    # partial circuit folding.
-    depth = len(circuit)
-    fractional_depth = int(depth * fractional_stretch / 2)
-    if fractional_depth != 0:
-        eye += inverse(circuit[-fractional_depth:]) + circuit[-fractional_depth:]
+    Parameters
+    ----------
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by.
+        finish_fold_method: A local folding method which is called to fold remaining
+                            gates to reach the input stretch factor, if necessary.
+        finish_fold_method_args: Any arguments to pass into finish_fold_method.
+    """
+
+    if not (stretch >= 1):
+        raise ValueError("The stretch factor must be a real number >= 1.")
 
-    return circuit + eye
+    if not circuit.are_all_measurements_terminal():
+        raise ValueError(f"Input circuit contains intermediate measurements and cannot be folded.")
+
+    folded = deepcopy(circuit)
+    measurements = _pop_measurements(folded)
+    base_circuit = deepcopy(folded)
+
+    # Determine the number of global folds and the final fractional stretch
+    num_global_folds = int(stretch // 3)
+    fractional_stretch = stretch % 3
+    ngates = len(list(circuit.all_operations()))
+    # Do the global folds
+    for _ in range(num_global_folds):
+        folded += Circuit(inverse(base_circuit), base_circuit)
+
+    # Do the local folds
+    # Adjust the stretch to account for the fact that there are now more gates in the circuit
+    folded_ngates = len(list(folded.all_operations()))
+    adjusted_stretch = 1. + ngates / folded_ngates * (fractional_stretch - 1)
+    if adjusted_stretch > 1:
+        folded = fold_local(folded, adjusted_stretch, finish_fold_method, finish_fold_method_args)
```

Comment:
In my opinion since we are folding the full circuit globally, it is better to fold globally also the final fraction of the circuit.
I.e. I think we should append something equivalent to the previous `inverse(circuit[-fractional_depth:]) + circuit[-fractional_depth:]`.  Which is different form folding individual gates locally. Note that folding gates locally would produce a discontinuous behavior when passing from `stretch`=2.999.. to `stretch`=3.

Reply:
Even with this strategy, wouldn't you still have the degree of freedom of which "part" of the unitary to fold globally at the end?

For example, suppose the unitary is `U = A B C` where each `A`, `B`, and `C` consists of equal numbers of gates. If we need to fold 1/3 of the gates at the end, is it clear which of the following we should do?

(1) `A B C --> A A^dag A B C` (fold globally from left)
(2) `A B C --> A B B^dag B C` (fold globally from middle)
(3) `A B C --> A B C C^dag C` (fold globally from right)

Reply:
Just for the sake of getting #42 merged, it might be good to keep the local folding methods for now and open an issue to discuss/address. I guess both local/global could be final folding methods, though maybe I'm over-complicating it.

Either way, this is an easy change which just entails something like

```python
ops = list(circuit.all_operations())
n = int(round(fractional_stretch * len(ops))
folded += cirq.Circuit([inverse(ops[-n:]), ops[-n:]])
# Or folded += cirq.Circuit([inverse(ops[:n]), ops[:n]]), etc.
```

I'll go with what you advise here @andreamari.

Reply:
Hi @rmlarose , sorry for my delayed reply. For global folding the natural choice is option (3), because ti is the only one which is consistent with integer folding. Let's make an example with more letters:  circuit = A B C D E.  A partial folding is something like this:
`(ABCDE) (E^dagD^dagC^dagB^dag)(BCDE)`,
now assume that we want to stretch a bit more, we would get:
`(ABCDE) (E^dagD^dagC^dagB^dagA^dag)(ABCDE) = circuit circuit^dag circuit,`
i.e. we converge to the correct integer circuit folding. This continuous stretching doesn't happen if we use different partial folding methods.

In conclusion I would vote for keeping it simple and not mixing local_folding with global_folding. However I also see that there is some urgency of merging. So, no problem, we can always fix this aspect in a new issue/PR.

Reply:
Thanks @andreamari, makes sense! Quick change so I just added it now.